src/lib/pubkey/classic_mceliece/cmce_matrix.cpp

```diff
@@ -0,0 +1,262 @@
+      // bit in the column of the current postion on the diagonal
+      // is still one
+      for(size_t row = 0; row < params.pk_no_rows(); ++row) {
+         if(row != diag_pos) {
+            mat[row].get().ct_conditional_xor(mat[row].at(diag_pos), mat[diag_pos].get());
+         }
+      }
+   }
+
+   return pivots;
+}
+
+std::vector<uint8_t> extract_pk_bytes_from_matrix(const Classic_McEliece_Parameters& params, const Matrix& mat) {
+   // Store T of the matrix (I_mt|T) as a linear vector to represent the
+   // public key as defined in McEliece ISO 9.2.7
+   std::vector<uint8_t> big_t(params.pk_size_bytes());
+   auto big_t_stuffer = BufferStuffer(big_t);
+
+   for(size_t row = 0; row < params.pk_no_rows(); ++row) {
+      auto pk_row_bits = mat[row].subvector(params.pk_no_rows());
+      auto row_bytes = pk_row_bits.to_bytes();
+
+      big_t_stuffer.append(row_bytes);
+   }
+
+   BOTAN_ASSERT_NOMSG(big_t_stuffer.full());
+
+   return big_t;
+}
+
+}  // namespace
+
+std::optional<std::pair<Classic_McEliece_Matrix, Column_Selection>> Classic_McEliece_Matrix::create_matrix(
+   const Classic_McEliece_Parameters& params,
+   const Classic_McEliece_Field_Ordering& field_ordering,
+   const Classic_McEliece_Minimal_Polynomial& g) {
+   auto mat = init_matrix_with_alphas(params, field_ordering, g);
+   auto pivots = apply_gauss(params, mat);
+
+   if(!pivots) {
+      return std::nullopt;
+   }
+
+   auto pk_mat_bytes = extract_pk_bytes_from_matrix(params, mat);
+   return std::make_pair(Classic_McEliece_Matrix(params, std::move(pk_mat_bytes)), pivots.value());
+}
+
+std::optional<std::pair<Classic_McEliece_Matrix, Column_Selection>>
+Classic_McEliece_Matrix::create_matrix_and_apply_pivots(const Classic_McEliece_Parameters& params,
+                                                        Classic_McEliece_Field_Ordering& field_ordering,
+                                                        const Classic_McEliece_Minimal_Polynomial& g) {
+   auto pk_matrix_and_pivots = create_matrix(params, field_ordering, g);
+
+   if(!pk_matrix_and_pivots) {
+      return std::nullopt;
+   }
+
+   auto& [_, pivots] = pk_matrix_and_pivots.value();
+
+   if(params.is_f()) {
+      field_ordering.permute_with_pivots(params, pivots);
+   }
+
+   return pk_matrix_and_pivots;
+}
+
+Code_Word Classic_McEliece_Matrix::mul(const Classic_McEliece_Parameters& params, const Error_Vector& e) const {
+   auto s = e.subvector(0, params.pk_no_rows());
+   auto e_T = e.subvector(params.pk_no_rows());
+   auto pk_slicer = BufferSlicer(m_mat_bytes);
+
+   for(size_t i = 0; i < params.pk_no_rows(); ++i) {
+      auto pk_current_bytes = pk_slicer.take(params.pk_row_size_bytes());
+      auto row = secure_bitvector(pk_current_bytes, params.n() - params.pk_no_rows());
+      row &= e_T;
+      s.at(i) = s.at(i) ^ row.has_odd_hamming_weight();
+   }
+
+   BOTAN_ASSERT_NOMSG(pk_slicer.empty());
+   return s.as<Code_Word>();
+}
```

Comment:
This produces quite a few copies and allocations. If that's in the implementation's hot path, we should probably want to have another look. Here's which (I believe cause allocations/copies):

1. `.subvector()` always creates a new `bitvector` and copies the content into a newly allocated buffer,
2. the c'tor of `bitvector` copies the passed-in buffer (particularly `pk_current_bytes`
3. `.as<>` produces a copy of the `bitvector` with a new type

For (3): This could already be the right type using `e.subvector<Code_Word>()`, no?

Reply:
Some context: This function is called once per encapsulation. Also, `e` and `s` are pretty small, while `m_mat_bytes` is gigantic. Therefore, 1) and 3) are not critical; I'll apply your suggestion for 3) anyway. I don't know how we can prevent 2) without having something like a bitvector view or dropping the bitvector altogether.

Reply:
I think I'll have a timeboxed look into a bitvector that doesn't own its underlying storage. Perhaps that isn't too hard to achieve, especially when we can limit it to byte-aligned subvectors.